File: Data Structures & Algorithms/buy-and-sell-crypto/submission-4.cpp
class Solution {
   public:
    int maxProfit(vector<int>& prices) {
        // brute force o(n^2)
        //     int profit = 0;
        //     for (int i = 0; i < prices.size(); i++) {
        //         for (int j = i+1; j < prices.size(); j++) {
        //             profit = max(profit, prices[j] - prices[i]);
        //         }
        //     }
        //     return profit;

        // two pointers
        int l = 0;
        int r = 1;
        int profit = 0;
        while (r < prices.size()) {
            if (prices[l] < prices[r])
                profit = max(profit, prices[r] - prices[l]);
            else if (prices[l] > prices[r])
                l = r;
            r += 1;
        }
        return profit;
    }
};

File: Data Structures & Algorithms/eating-bananas/submission-4.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        # brute force
        # speed = 1
        # while True:
        #     totalTime = 0 
        #     for p in piles:
        #         totalTime += math.ceil(p / speed)
            
        #     if totalTime <= h:
        #         return speed
            
        #     speed += 1 # Try the next speed

        l, r = 1, max(piles)

        totalHours = 0
        res = max(piles)

        while l<=r:
            k = l + ((r-l)//2)
            totalHours=0
            for p in piles:
                totalHours += math.ceil(p/k)
            if totalHours <= h:
                res = min(res,k)
                r = k-1
            else:
                l = k+1 
        return res